src/workerd/jsg/async-context.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include "jsg.h"
+#include <v8.h>
+
+namespace workerd::jsg {
+
+class AsyncContextFrame: public kj::Refcounted {
+  // Provides for basic internal async context tracking. Eventually, it is expected that
+  // this will be provided by V8 assuming that the AsyncContext proposal advances through
+  // TC-39. For now, however, we implement a model that is similar but not quite identical
+  // to that implemented by Node.js.
+  //
+  // At any point in time when JavaScript is running, there is a current "Async Context Frame",
+  // within which any number of "async resources" can be created. The term "resource" here
+  // comes from Node.js (which really doesn't take the time to define it properly). Conceptually,
+  // an "async resource" is some Thing that generates asynchronous activity over time (either
+  // once or repeatedly). For instance, a timer is an async resource that invokes a callback
+  // after a certain period of time elapses; a promise is an async resource that may trigger
+  // scheduling of a microtask at some point in the future, and so forth. Whether or not
+  // "resource" is the best term to use to describe these, it's what we have because our
+  // intent here is to stay aligned with Node.js' model as closely as possible.
+  //
+  // Every async resource maintains a reference to the Async Context Frame that was current
+  // at the moment the resource is created.
+  //
+  // Frames form a stack. The default frame is the Root. We "enter" a frame by pushing it
+  // onto to top of the stack (making it "current"), then perform some action within that
+  // frame, then "exit" by popping it back off the stack. The Root is associated with the
+  // Isolate itself such that every isolate always has at least one frame on the stack at
+  // all times. In Node.js terms, the "Async Context Frame" would be most closely aligned
+  // with the concept of an "execution context" or "execution scope".
+  //
+  // Every Frame has a storage context. The current frame determines the currently active
+  // storage context. So, for instance, when we start executing, the Root Frame's storage
+  // context is active. When a timeout elapses and a timer is going to fire, we enter the
+  // timer's Frame which makes that frame's storage context active. Once the timer
+  // callback has completed, we return back to the Root frame and storage context.
+  //
+  // All frames (except for the Root) are created within the scope of a parent, which by
+  // default is whichever frame is current when the new frame is created. When the new frame
+  // is created, it inherits a copy storage context of the parent.
+public:
+  class StorageKey: public kj::Refcounted {
+    // An opaque key that identifies an async-local storage cell within the frame.
+  public:
+    StorageKey() : hash(kj::hashCode(this)) {}
+
+    void reset() { dead = true; }
+    // The owner of the key should reset it when it goes away.
```

Comment:
Could use a little more commentary on when this is expected to be used. It feels a bit non-obvious to me.

I guess this is when the AsyncLocalStorage itself is GC'd, which implies that this slot in the map can never be looked up again so might as well be dropped? That said, ALS objects are typically globals with unlimited lifetime, right? So this is not commonly used?

I kind of wonder if we should be using a WeakMap. Otherwise you could have an uncollectable cycle-like situation where someone has stored an ALS value that holds a reference back to its key, but the key is not otherwise reachable so will never be used. That said it seems like an unusual use case, so could be left as a TODO.

Reply:
Yeah, this bit is a little wonky. A WeakMap would be ideal but v8 doesn't currently give us a C++ API to create a WeakMap unless I've missed it somewhere. I'll extend the comment here a bit.